simd/src/Kokkos_Simd_Common.hpp

```diff
@@ -0,0 +1,262 @@
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_SIMD_COMMON_HPP
+#define KOKKOS_SIMD_COMMON_HPP
+
+#include <cmath>
+#include <cstring>
+
+#include <Kokkos_Core.hpp>
+
+namespace Kokkos {
+namespace Experimental {
+
+template <class To, class From>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION constexpr To bit_cast(
+    From const& src) {
+  To dst;
+  std::memcpy(&dst, &src, sizeof(To));
+  return dst;
+}
+
+template <class T, class Abi>
+class simd;
+
+template <class T, class Abi>
+class simd_mask;
+
+struct element_aligned_tag {};
+
+template <class Mask, class Value>
+class const_where_expression {
+ protected:
+  Value& m_value;
+  Mask const& m_mask;
+
+ public:
+  const_where_expression(Mask const& mask_arg, Value const& value_arg)
+      : m_value(const_cast<Value&>(value_arg)), m_mask(mask_arg) {}
+};
+
+template <class Mask, class Value>
+class where_expression : public const_where_expression<Mask, Value> {
+  using base_type = const_where_expression<Mask, Value>;
+
+ public:
+  where_expression(Mask const& mask_arg, Value& value_arg)
+      : base_type(mask_arg, value_arg) {}
+};
+
+template <class T, class Abi>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION
+    where_expression<simd_mask<T, Abi>, simd<T, Abi>>
+    where(Kokkos::Impl::identity_t<simd_mask<T, Abi>> const& mask,
+          simd<T, Abi>& value) {
+  return where_expression(mask, value);
+}
+
+template <class T, class Abi>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION
+    const_where_expression<simd_mask<T, Abi>, simd<T, Abi>>
+    where(Kokkos::Impl::identity_t<simd_mask<T, Abi>> const& mask,
+          simd<T, Abi> const& value) {
+  return const_where_expression(mask, value);
+}
+
+template <class T, class Abi>
+KOKKOS_FORCEINLINE_FUNCTION simd<T, Abi>& operator+=(
+    simd<T, Abi>& a, Kokkos::Impl::identity_t<simd<T, Abi>> const& b) {
+  a = a + b;
+  return a;
+}
+
+template <class T, class Abi>
+KOKKOS_FORCEINLINE_FUNCTION simd<T, Abi>& operator-=(
+    simd<T, Abi>& a, Kokkos::Impl::identity_t<simd<T, Abi>> const& b) {
+  a = a - b;
+  return a;
+}
+
+template <class T, class Abi>
+KOKKOS_FORCEINLINE_FUNCTION simd<T, Abi>& operator*=(
+    simd<T, Abi>& a, Kokkos::Impl::identity_t<simd<T, Abi>> const& b) {
+  a = a * b;
+  return a;
+}
+
+template <class T, class Abi>
+KOKKOS_FORCEINLINE_FUNCTION simd<T, Abi>& operator/=(
+    simd<T, Abi>& a, Kokkos::Impl::identity_t<simd<T, Abi>> const& b) {
+  a = a / b;
+  return a;
+}
+
+template <class U, class T, class Abi>
+[[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION
+    std::enable_if_t<std::is_arithmetic_v<U>, simd<T, Abi>>
+    operator+(U const& a, simd<T, Abi> const& b) {
+  return simd<T, Abi>(T(a)) + b;
+}
```

Comment:
I think this was to emulate arithmetic promotions since the compiler doesn't twice-implicit-cast on its own. For example:
```cpp
auto minus_42 = -1 * simd<double,native>(42.0);
```
Where the left hand side is type `int`... I'm not 100% sure anymore though.

Reply:
Yep, I just removed this to confirm, and I could no longer add `int` to a `simd<double,...>`.

Reply:
Your original comment was correct, I made it work the way it is supposed to now